### PR TITLE
Enabled JSON and non-JSON handling of Dates and Times

### PR DESCRIPTION
**Pull request by brendandixon**
Signed-off-by: Brendan Dixon <brendand@microsoft.com>

### CODE REVIEW

File: autorest/date/date.go [hunk 1]
```diff
@@ -28,27 +30,37 @@ type Date struct {
...
+	return d, err
 }
 
 // ByUnmarshallingDate returns a RespondDecorator that decodes the http.Response Body into a Date
 // pointed to by d.
 func ByUnmarshallingDate(d *Date) autorest.RespondDecorator {
+	return byUnmarshallingDate(d, rfc3339FullDate)
+}
+
+// ByUnmarshallingjsonDate returns a RespondDecorator that decodes JSON within the http.Response
```
Comment (ahmetalpbalkan) [inline autorest/date/date.go]: ByUnmarshallingJSONDate

File: autorest/date/date.go [hunk 2]
```diff
@@ -28,27 +30,37 @@ type Date struct {
...
-func readDate(r io.Reader) (Date, error) {
-	b, err := ioutil.ReadAll(r)
-	if err == nil {
-		return ParseDate(string(b))
-	}
-	return Date{}, err
+func parseDate(date string, format string) (d Date, err error) {
+	d = Date{}
+	d.Time, err = time.Parse(format, date)
+	return d, err
```
Comment (ahmetalpbalkan) [inline autorest/date/date.go]: you can make it even shorter if you like

```go
	d, err = time.Parse(format, date)
	return Date{Time:d}, err
 ```

Comment (ahmetalpbalkan) [inline autorest/date/date.go]: and this way you can get rid of naming `d` and `err` in the signature.

File: autorest/date/time_test.go [hunk 3]
```diff
@@ -181,27 +181,59 @@ func TestDate_ByUnmarshallingTime(t *testing.T) {
...
+
+	r := mocks.NewResponseWithContent("2001-02-03T04:05:06Z")
+	err := autorest.Respond(r,
+		byUnmarshallingTime(&t1, rfc3339),
+		autorest.ByClosing())
+	if err != nil {
+		t.Errorf("date: byUnmarshallingTime failed (%v)", err)
+	}
+	if !reflect.DeepEqual(t1, t2) {
+		t.Errorf("date: byUnmarshallingTime failed to properly unmarshall -- expected %v, received %v", t2, t1)
```
Comment (ahmetalpbalkan) [inline autorest/date/time_test.go]: you can use `%#v` for these. (same comment for the test above)

**Comment by ahmetalpbalkan** [comment]
lgtm

File: autorest/date/date.go (hunk 2 above)
Comment (brendandixon) [inline autorest/date/date.go]: @ahmetalpbalkan Done.

**Pull request merged**